File: tests/ten_runtime/integration/go/expired_ten_go/expired_ten_go_app/ten_packages/extension/extension_a/extension.go
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//

package default_extension_go

import (
	"fmt"
	"time"

	"go_common_dep/types"
	ten "ten_framework/ten_runtime"
)

type extensionA struct {
	ten.DefaultExtension
}

func newExtensionA(name string) ten.Extension {
	return &extensionA{}
}

func (p *extensionA) OnCmd(
	tenEnv ten.TenEnv,
	cmd ten.Cmd,
) {
	go func() {
		fmt.Println("extensionA OnCmd")

		cmdB, _ := ten.NewCmd("B")
		tenEnv.SendCmd(cmdB, func(r ten.TenEnv, cs ten.CmdResult, e error) {
			detail, err := cs.GetPropertyPtr("data")
			if err != nil {
				cmdResult, _ := ten.NewCmdResult(ten.StatusCodeError, cmd)
				cmdResult.SetPropertyString("detail", err.Error())
				tenEnv.ReturnResult(cmdResult, nil)
			} else {
				cmdResult, _ := ten.NewCmdResult(ten.StatusCodeOk, cmd)
				cmdResult.SetPropertyString("detail", detail.(*types.UserData).Name)
				tenEnv.ReturnResult(cmdResult, nil)
			}
		})

		time.Sleep(time.Second * 3)

		// Use the expired tenEnv object.
		newCmdB, _ := ten.NewCmd("B")
		err := tenEnv.SendCmd(newCmdB, nil)
		if err != nil {
			fmt.Println("Failed to use invalid ten object.")
		} else {
			panic("should not happen")
		}
	}()
}

func init() {
	// Register addon
	err := ten.RegisterAddonAsExtension(
		"extension_a",
		ten.NewDefaultExtensionAddon(newExtensionA),
	)
	if err != nil {
		fmt.Println("register addon failed", err)
	}
}

File: tests/ten_runtime/integration/python/start_graph_memory_leak_python/start_graph_memory_leak_python_app/ten_packages/extension/test_extension_2/extension.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#

from ten_runtime import (
    Extension,
    TenEnv,
    LogLevel,
)


class TestExtension2(Extension):
    """A simple extension used as a subgraph node for memory leak testing.

    This extension does nothing special - it just starts and stops normally,
    allowing the test to measure whether repeated graph creation/destruction
    leaks memory.
    """

    def __init__(self, name: str) -> None:
        super().__init__(name)

    def on_start(self, ten_env: TenEnv) -> None:
        ten_env.log(LogLevel.DEBUG, "on_start")
        ten_env.on_start_done()

    def on_stop(self, ten_env: TenEnv) -> None:
        ten_env.log(LogLevel.DEBUG, "on_stop")
        ten_env.on_stop_done()
